fusion.ml

```diff
@@ -313,49 +327,48 @@ let rec type_subst i ty =
 
   (*This is used when checking quote types match the term types as holes should always be of type epsilon - type_of returns the type of the thing inside the quote so that they can be used more easily
   in the parser*)
-  let rec qcheck_type_of tm = match tm with
-      Var(_,ty) -> ty
+  let rec qcheck_type_of = function
+    | Var(_,ty) -> ty
     | Const(_,ty) -> ty
-    | Comb(s,_) -> (match qcheck_type_of s with Tyapp("fun",[dty;rty]) -> rty | Tyapp("epsilon",[]) -> Tyapp("epsilon",[]))
+    | Comb(s,_) -> (match qcheck_type_of s with Tyapp("fun",[dty;rty]) -> rty | ep_ty -> ep_ty)
     | Abs(Var(_,ty),t) -> Tyapp("fun",[ty;qcheck_type_of t])
-    | Quote(e,_) -> Tyapp("epsilon",[])
+    | Quote(e,_) -> ep_ty
     | Hole(e,ty) -> ty
     | Eval(e,ty) -> ty
     | _ -> failwith "TYPE_OF: Invalid term. You should not see this error under normal use, if you do, the parser has allowed an ill formed term to be created."
 
-  let rec type_of tm =
-    match tm with
-      Var(_,ty) -> ty
+  let rec type_of = function
+    | Var(_,ty) -> ty
     | Const(_,ty) -> ty
-    | Comb(s,_) -> (match type_of s with Tyapp("fun",[dty;rty]) -> rty| Tyapp("epsilon",[]) -> Tyapp("epsilon",[]))
+    | Comb(s,_) -> (match type_of s with Tyapp("fun",[dty;rty]) -> rty| ep_ty -> ep_ty)
     | Abs(Var(_,ty),t) -> Tyapp("fun",[ty;type_of t])
-    | Quote(e,_) -> Tyapp("epsilon",[])
+    | Quote(e,_) -> ep_ty
     | Hole(e,ty) -> ty
     | Eval(e,ty) -> ty
     | _ -> failwith "TYPE_OF: Invalid term. You should not see this error under normal use, if you do, the parser has allowed an ill formed term to be created."
 
   (*Internal function to grab the type of an applied function*)
   let fun_type_of tm = 
-    let rec ftype_of trm = match trm with
+    let rec ftype_of = function
     | Comb(l,_) -> ftype_of l
     | Const(n,t) | Var(n,t) when not (is_vartype t) && fst (dest_type t) = "fun" -> t  
     | _ -> failwith "Not a function"
-
   in
-
   match tm with
-    | Comb(l,r) when type_of tm = Tyapp("epsilon",[]) -> ftype_of l 
-    | _ -> failwith "Incomplete or mistyped function" 
+    | Comb(l,r) when type_of tm = ep_ty -> ftype_of l 
+    | _ -> failwith "Incomplete or mistyped function"   
 
   (*Checks if a term is eval-free*)
-  let rec is_eval_free tm = match tm with
-    | Var(_,_) -> true
-    | Const(_,_) -> true
-    | Comb(a,b) -> is_eval_free a && is_eval_free b
-    | Abs(a,b) -> is_eval_free a && is_eval_free b
+  let rec is_eval_free tm = match tm with 
+    | Var(_,_) -> Ok
+    | Const(_,_) -> Ok
+    | Comb(a,b) -> if is_eval_free a = Ok then is_eval_free b else is_eval_free a 
```

Comment:
You want to let-bind `is_eval_free a` so that you can re-use it in the `else` clause.  [If Ocaml had monads, it would be even simpler, but...]